hexa/plugins/connector_gcs/models.py

```diff
@@ -0,0 +1,174 @@
+        verbose_name = "GCS Credentials"
+        verbose_name_plural = "GCS Credentials"
+        ordering = ("service_account",)
+
+    service_account = models.CharField(max_length=30)
+    project_id = models.CharField(max_length=30)
+    client_id = models.CharField(max_length=30)
+    client_email = models.CharField(max_length=100)
+    client_x509_cert_url = models.CharField(max_length=150)
+    auth_uri = models.CharField(
+        max_length=100, default="https://accounts.google.com/o/oauth2/auth"
+    )
+    token_uri = models.CharField(
+        max_length=100, default="https://oauth2.googleapis.com/token"
+    )
+    auth_provider_x509_cert_url = models.CharField(
+        max_length=100, default="https://www.googleapis.com/oauth2/v1/certs"
+    )
+
+    private_key_id = EncryptedTextField(max_length=50)
+    private_key = EncryptedTextField()
+
+    @property
+    def display_name(self):
+        return self.service_account
+
+
+class BucketQuerySet(BaseQuerySet):
+    def filter_for_user(
+        self,
+        user: typing.Union[AnonymousUser, user_management_models.User],
+        mode: PermissionMode = None,
+        mode__in: typing.Sequence[PermissionMode] = None,
+    ):
+        if mode is not None and mode__in is not None:
+            raise ValueError('Please provide either "mode" or "mode_in" - not both')
+
+        if user.is_superuser or user.has_feature_flag("gcs"):
+            return self
+        else:
+            return self.none()
+
+
+class Bucket(Datasource):
+    def get_permission_set(self):
+        return self.gcsbucketpermission_set.all()
+
+    class Meta:
+        verbose_name = "GCS Bucket"
+        ordering = ("name",)
+
+    name = models.CharField(max_length=200)
+
+    objects = BucketQuerySet.as_manager()
+    searchable = True
+
+    @property
+    def principal_credentials(self):
+        try:
+            return Credentials.objects.get()
+        except (Credentials.DoesNotExist, Credentials.MultipleObjectsReturned):
+            raise ValidationError(
+                "The GCS connector plugin should be configured with a single Credentials entry"
+            )
+
+    @property
+    def content_summary(self):
+        count = 0
+
+        return (
+            ""
+            if count == 0
+            else _("%(count)d object%(suffix)s")
+            % {"count": count, "suffix": pluralize(count)}
+        )
+
+    def populate_index(self, index):
+        index.last_synced_at = self.last_synced_at
+        index.content = self.content_summary
+        index.path = [self.pk.hex]
+        index.external_id = self.name
+        index.external_name = self.name
+        index.external_type = "bucket"
+        index.search = f"{self.name}"
+        index.datasource_name = self.name
+        index.datasource_id = self.id
+
+    @property
+    def display_name(self):
+        return self.name
+
+    def __str__(self):
+        return self.display_name
+
+    def writable_by(self, user):
```

Comment:
I know that this is what we have for S3 as well, but could you try using the new permissions for this part? And maybe adapt the S3 code as well?